src/array/cuda/gather_mm.cu

```diff
@@ -0,0 +1,334 @@
+    const DType* B, int ldb, const DType* beta,
+    DType* C, int ldc) {
+  LOG(INFO) << "Not supported dtype";
+  return CUBLAS_STATUS_EXECUTION_FAILED;
+}
+
+template <>
+cublasStatus_t cublasGemm<float>(cublasHandle_t handle, cublasOperation_t transa,
+    cublasOperation_t transb, int m, int n, int k,
+    const float* alpha, const float* A, int lda,
+    const float* B, int ldb, const float* beta,
+    float* C, int ldc) {
+  return cublasSgemm(handle, transa, transb, m, n, k, alpha, A, lda,
+      B, ldb, beta, C, ldc);
+}
+
+template <>
+cublasStatus_t cublasGemm<double>(cublasHandle_t handle, cublasOperation_t transa,
+    cublasOperation_t transb, int m, int n, int k,
+    const double* alpha, const double* A, int lda,
+    const double* B, int ldb, const double* beta,
+    double* C, int ldc) {
+  return cublasDgemm(handle, transa, transb, m, n, k, alpha, A, lda,
+      B, ldb, beta, C, ldc);
+}
+
+/*
+ * \brief Tranpose the input matrix.
+ * \param row number of rows of input matrix.
+ * \param col number of columns of input matrix.
+ */
+template <typename DType>
+void _Transpose(cublasHandle_t handle,
+                const DType* in, DType* out,
+                int row, int col) {
+  DType alpha = 1., beta = 0.;
+  CUBLAS_CALL(Xgeam<DType>(
+      handle,
+      CUBLAS_OP_T,
+      CUBLAS_OP_N,
+      row, col,
+      &alpha, in, col,
+      &beta, nullptr, row,
+      out, row));
+}
+
+}  // namespace
+
+/* \Note
+   Idea 1: tranpose B and compute dot product of each row vector of A
+   and column vector of B. Reuse in A.
+   Idea 2: multiply 1 element of A with a row of B and compute partial
+   result of the output
+*/
+
+/* Implementation of Idea 2
+  \Note One warp is assigned to process one row of A. Each WARP sequentially multiplies
+  one element of A and a row of B to compute partial result of the output. A is
+  loaded in shared memory in a coalesced way. Output matrix is loaded in registers.
+  B should get benefit from L2 cache.
+*/
+
+template <typename Idx, typename DType>
+__global__ void gatherMMUnsortedEKernel(
+    const DType* __restrict__ A,
+    const DType* __restrict__ B,
+    DType* __restrict__ C,
+    const Idx* __restrict__ etype,
+    int64_t num_rows,
+    int64_t k_start, int64_t in_len, int64_t out_len) {
+    unsigned int tId = threadIdx.x;
+    unsigned int laneId = tId & 31;
+    unsigned int gId = (blockIdx.x * blockDim.x + threadIdx.x);
+    unsigned int warpId = gId >> 5;
+    unsigned int row = warpId;
+    if (row < num_rows) {
+        unsigned int local_row = row & 3;  // hardcoded for TB size 128 (4 warps)
+        const int sh_h_tile = 64;
+        __shared__ DType sh_H[4 * sh_h_tile];
+        int h_tile = sh_h_tile;
+        if ((in_len - k_start) < h_tile) h_tile = in_len - k_start;
+        /* Load A in shared mem in a coalesced way */
+        for (unsigned int l = laneId; l < h_tile; l += 32)
+            sh_H[local_row * h_tile + l] = A[row * in_len + (k_start + l)];
```

Comment:
We should probably keep indexing into `sh_H` fixed, so as to avoid weird performance behavior related to memory bank conflicts as `h_tile` changes:
```
sh_H[local_row * sh_h_tile +l]
```